include/odp/api/spec/packet_io.h

```diff
@@ -256,11 +256,27 @@ typedef struct odp_pktio_param_t {
  * belong to time synchronization protocol (PTP).
  *
  * Packet input checksum checking may be enabled or disabled. When it is
- * enabled, implementation will verify checksum correctness on incoming packets
- * and depending on drop configuration either deliver erroneous packets with
- * appropriate flags set (e.g. odp_packet_has_l3_error()) or drop those.
- * When packet dropping is enabled, application will never receive a packet
- * with the specified error and may avoid to check the error flag.
+ * enabled, implementation will attempt to verify checksum correctness on
+ * incoming packets and depending on drop configuration either deliver erroneous
+ * packets with appropriate flags set (e.g. odp_packet_has_l3_error(),
+ * odp_packet_l3_chksum_status()) or drop those. When packet dropping is
+ * enabled, application will never receive a packet with the specified error
+ * and may avoid to check the error flag.
+ *
+ * If checksum checking is enabled, IPv4 header checksum checking is always
+ * done for packets that do not have IP options and L4 checksum checking
+ * is done for unfragmented packets that do not have IPv4 options or IPv6
+ * extension headers. In other cases checksum checking may or may not
+ * be done. For example, L4 checksum of fragmented packets is typically
+ * not checked.
```

Comment:
L4 checksums being skipped for IP fragments is normal, but why the caveat about IP options or IPv6 extension headers? A TCP or UCP checksum is well defined independent of these L3 considerations.

Reply:
L4 check includes fields from IP header. If the IP header is not in basic form (has options/extensions) HW may not be able to do the L4 checksum check.

Reply:
Again, TCP checksums are not optional and their presence is not conditional upon such things. If the intent here is to request best-effort HW offload (as opposed to requiring the ODP implementation provide this service however it chooses) then this wording needs to change to reflect that distinction. I'd take the position that applications should never have to worry about checksums unless they are being overridden for specific diagnostic or other purposes. In normal operation applications should be able to rely on the ODP implementation taking care of them (preferably in HW, but otherwise in SW if needed).

Reply:
The intention is best effort on top of the minimum requirement that all basic IP packets must be checked. The wording says that quite clearly already (what needs to change ?). Again it's a fact that not all HW will check non-basic IP packets, or if they do (e.g. with firmware / SW) packet rate may be much lower for _all_ packets.

Typically, these non-basic packets are a small fraction of all packets and belong to the slow path anyway.

Reply:
It would not be good to require ODP to always check the L4 checksum if checksum checking is enabled in the config, because then some implementations might have to fall back to SW checking for some packets and doing the check in SW would be a waste of resources in case the application is not interested in the L4 checksum of all packets (e.g. an application that both terminates and forwards IP packets needs to check the L4 checksum of the locally received packets but should ignore the L4 checksum of forwarded packets.

Reply:
@JannePeltonen I thought that's why we were providing the per-packet override capability here--so that applications can say "While I normally want checksums added, for this packet that's not needed". Or "Please don't add checksums by default, however for this packet please do so". The point is, if the application determines that checksum processing is needed for a particular packet (or interface) then that work needs to be done. The ODP implementation is in a better position than the application to determine how best to do that work on this platform since it has direct access to the platform-specific HW, specialized instructions, etc., and if it needs to fall back to SW can do so using instruction sequences optimized to its microarchitecture.

On the RX side, again we either don't care about checksum validation or else we do. If we do then that work has to be done. So I don't see what advantage is gained by requiring every application to have a "backup plan" in this area.

Reply:
In the RX side there are valid use cases where we both care and do not care about the L4 checksum. If a system is both forwarding and terminating IP traffic and neither clearly dominates the other, then we want to be able to take advantage of HW checksum checking without unnecessarily doing SW checksum checking when it is not needed.

ODP implementation cannot determine which packets are going to be forwarded (and for which checking the L4 checksum would be a waste of resources) and which are locally received (and L4 checksum must be checked). The application knows it.

It would be good if an application would be able to request "cheap" (i.e. done in HW, not in SW) L4 checksum checking for all incoming packets and later decide whether the result of the checksum check is actually used and whether to fall back to the much more costly SW calculation (once it is known for which packets the check is needed.

Maybe the API proposal is not clear enough of the intention of enabling this, but this is what we have needed the past in actual products and this is what we have implemented (in totally different context than ODP).

Maybe one possibility would be to have ODP always check the L4 checksum but do the SW fallback only when the checksum result is being requested. If the application checks the L4 checksum result only when it knows that it needs to check it, the SW calculation would not be done in vain for other packets.

Even for applications that only terminate all traffic and thus need to check the L4 checksum, it would make sense to postpone the SW fallback until it is known that the checksum result is really needed (and the packet did not fail any other checks, i.e. had proper L3 checksum, the right IP addresses and ports etc).

Reply:
The problem of lazy parsing (incl checksum) is that application may alter the packet before asking something about it. Parser results reflect the packet format when it first arrived the system.

Agree with Janne that cheap checksum status check is needed to avoid ODP implementation to do extra work for nothing.